### PR TITLE
remove support for rustc_safe_intrinsic attribute; use rustc_intrinsic functions instead

### PR DESCRIPTION
**Pull request by RalfJung**
This brings us one step closer towards removing support for `extern "rust-intrinsic"` blocks, in favor of `#[rustc_intrinsic]` functions.

Also move `#[rustc_intrinsic]` under the `intrinsics` feature gate, to match the `extern "rust-intrinsic"` style.

### CODE REVIEW

**Comment by RalfJung** [comment]
@antoyo @GuillaumeGomez updating the GCC tests here was unnecessarily painful, since the same code using very unstable features is duplicated in 12 files. Would be nice if that could be reduced -- can't they all use mini_core?

**Comment by RalfJung** [comment]
Oh great, gcc and cranelift have different versions of the mini_core file, so one can't even just copy-paste things from one to the other. :(

**Comment by antoyo** [comment]
> Oh great, gcc and cranelift have different versions of the mini_core file, so one can't even just copy-paste things from one to the other. :(

If the one from cg_cranelift works for the GCC codegen, you can just copy it.

**Comment by antoyo** [comment]
> @antoyo @GuillaumeGomez updating the GCC tests here was unnecessarily painful, since the same code using very unstable features is duplicated in 12 files. Would be nice if that could be reduced -- can't they all use mini_core?

Good idea. I opened [an issue](https://github.com/rust-lang/rustc_codegen_gcc/issues/571).

File: compiler/rustc_middle/src/ty/util.rs [hunk 1]
```diff
@@ -1785,9 +1785,9 @@ pub fn is_doc_notable_trait(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
 /// the compiler to make some assumptions about its shape; if the user doesn't use a feature gate, they may
 /// cause an ICE that we otherwise may want to prevent.
 pub fn intrinsic_raw(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<ty::IntrinsicDef> {
-    if (matches!(tcx.fn_sig(def_id).skip_binder().abi(), ExternAbi::RustIntrinsic)
-        && tcx.features().intrinsics())
-        || (tcx.has_attr(def_id, sym::rustc_intrinsic) && tcx.features().rustc_attrs())
+    if tcx.features().intrinsics()
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply compiler/rustc_middle/src/ty/util.rs:1788]: yeah that's what I meant to write, forgot to dully clean up the second arm... *oops*

**Review by compiler-errors — APPROVED** [review]

**Comment by workingjubilee** [comment]
seems conflicty.

@bors p=1

**Comment by workingjubilee** [comment]
@RalfJung sorry, I think it wasn't actually necessary to r- that but bors said there were and ??? and I didn't want to wait to find out if the commit proved dubious after it landed.

**Comment by RalfJung** [comment]
I rebased, hopefully that makes bors reconsider its choices.

@bors r=compiler-errors

**Comment by Mark-Simulacrum** [comment]
Dropping regression label, appears to be almost certainly bimodality in benchmarks (i.e., noise for triage purposes).

**Pull request merged**